File: openpifpaf/decoder/nms.py
import logging
import time

import numpy as np

from .occupancy import Occupancy

LOG = logging.getLogger(__name__)


class Keypoints:
    suppression = 0.0
    instance_threshold = 0.0
    keypoint_threshold = 0.0
    occupancy_visualizer = None

    def annotations(self, anns):
        start = time.perf_counter()

        for ann in anns:
            ann.data[ann.data[:, 2] < self.keypoint_threshold] = 0.0
        anns = [ann for ann in anns if ann.score() >= self.instance_threshold]

        if not anns:
            return anns

        occupied = Occupancy((
            len(anns[0].data),
            int(max(np.max(ann.data[:, 1]) for ann in anns) + 1),
            int(max(np.max(ann.data[:, 0]) for ann in anns) + 1),
        ), 2, min_scale=4)

        anns = sorted(anns, key=lambda a: -a.score())
        for ann in anns:
            assert ann.joint_scales is not None
            assert len(occupied) == len(ann.data)
            for f, (xyv, joint_s) in enumerate(zip(ann.data, ann.joint_scales)):
                v = xyv[2]
                if v == 0.0:
                    continue

                if occupied.get(f, xyv[0], xyv[1]):
                    xyv[2] *= self.suppression
                else:
                    occupied.set(f, xyv[0], xyv[1], joint_s)  # joint_s = 2 * sigma

        if self.occupancy_visualizer is not None:
            LOG.debug('Occupied fields after NMS')
            self.occupancy_visualizer.predicted(occupied)

        for ann in anns:
            ann.data[ann.data[:, 2] < self.keypoint_threshold] = 0.0
        anns = [ann for ann in anns if ann.score() >= self.instance_threshold]
        anns = sorted(anns, key=lambda a: -a.score())

        LOG.debug('nms = %.3f', time.perf_counter() - start)
        return anns


class Detection:
    suppression = 0.1
    suppression_soft = 0.3
    instance_threshold = 0.1
    iou_threshold = 0.7
    iou_threshold_soft = 0.5

    @staticmethod
    def bbox_iou(box, other_boxes):
        box = np.expand_dims(box, 0)
        x1 = np.maximum(box[:, 0], other_boxes[:, 0])
        y1 = np.maximum(box[:, 1], other_boxes[:, 1])
        x2 = np.minimum(box[:, 0] + box[:, 2], other_boxes[:, 0] + other_boxes[:, 2])
        y2 = np.minimum(box[:, 1] + box[:, 3], other_boxes[:, 1] + other_boxes[:, 3])
        inter_area = np.maximum(0.0, x2 - x1) * np.maximum(0.0, y2 - y1)
        box_area = box[:, 2] * box[:, 3]
        other_areas = other_boxes[:, 2] * other_boxes[:, 3]
        return inter_area / (box_area + other_areas - inter_area + 1e-5)

    def annotations(self, anns):
        start = time.perf_counter()

        anns = [ann for ann in anns if ann.score >= self.instance_threshold]
        if not anns:
            return anns
        anns = sorted(anns, key=lambda a: -a.score)

        all_boxes = np.stack([ann.bbox for ann in anns])
        for ann_i, ann in enumerate(anns[1:], start=1):
            mask = [ann.score >= self.instance_threshold for ann in anns[:ann_i]]
            ious = self.bbox_iou(ann.bbox, all_boxes[:ann_i][mask])
            max_iou = np.max(ious)

            if max_iou > self.iou_threshold:
                ann.score *= self.suppression
            elif max_iou > self.iou_threshold_soft:
                ann.score *= self.suppression_soft

        anns = [ann for ann in anns if ann.score >= self.instance_threshold]
        anns = sorted(anns, key=lambda a: -a.score)

        LOG.debug('nms = %.3f', time.perf_counter() - start)
        return anns

File: cpp/examples/cifcafdecoder_v0.12.13_wholebody/src/decoder/nms.cpp
#include <algorithm>
#include <cassert>

#include "../../include/decoder/annotation.hpp"
#include "../../include/decoder/occupancy.hpp"
#include "../../include/utils/numpy_utils.hpp"
#include "../../include/decoder/nms.hpp"


bool compare_annotations(Annotation &a1, Annotation &a2) {
    return a1.score() > a2.score();
}

vector<Annotation> Keypoints::annotations(vector<Annotation> &anns) {
    for(int i = 0; i < anns.size(); i++) {
        vector<float> v;
        for(int j = 0; j < anns[i].data.size(); j++) {
            v.push_back(anns[i].data[j][2]);
        }
        vector<bool> mask = mask1d_lt(v, this->keypoint_threshold);

        for(int k = 0; k < mask.size(); k++) {
            if(mask[k]) {
                for(int l = 0; l < anns[i].data[0].size(); l++) {
                    anns[i].data[k][l] = 0.0;
                }
            }
        }
    }

    vector<Annotation> filtered_annotations;
    for(int i = 0; i < anns.size(); i++) {
        if(anns[i].score() >= this->instance_threshold)
            filtered_annotations.push_back(anns[i]);
    }
    if(filtered_annotations.empty())
        return filtered_annotations;

    int d1 = filtered_annotations[0].data.size();
    float d2;
    float d3;
    for(int i = 0; i < filtered_annotations.size(); i++) {
        vector<float> v2;
        vector<float> v3;
        for(int j = 0; j <  filtered_annotations[i].data.size(); j++) {
            v2.push_back(filtered_annotations[i].data[j][1]);
            v3.push_back(filtered_annotations[i].data[j][0]);
        }
        float max_v2 = maximum1d(v2);
        float max_v3 = maximum1d(v3);
        if(i == 0) {
            d2 = max_v2;
            d3 = max_v3;
        } else {
            if(max_v2 > d2)
                d2 = max_v2;
            if(max_v3 > d3)
                d3 = max_v3;
        }
    }

    vector<int> shape = {d1, (int)(d2 + 1), (int)(d3 + 1)};
    Occupancy occupied = Occupancy(shape, 2.0, 4.0);

    sort(filtered_annotations.begin(), filtered_annotations.end(), compare_annotations);
    for(int i = 0; i < filtered_annotations.size(); i++) {
        assert(!filtered_annotations[i].joint_scales.empty() && "Annotation join scales must not be empty");
        assert(occupied.occupancy.size() == filtered_annotations[i].data.size() && "Occupied and annotation data size must match");

        for(int j = 0; j < filtered_annotations[i].data.size(); j++) {
            vector<float> xyv = filtered_annotations[i].data[j];
            float joint_s = filtered_annotations[i].joint_scales[j];
            float v = xyv[2];
            if(v == 0.0)
                continue;

            if(occupied.get(j, xyv[0], xyv[1])) {
                xyv[2] *= this->suppression;
            } else {
                occupied.set(j, xyv[0], xyv[1], joint_s);
            }
        }
    }

    for(int i = 0; i < filtered_annotations.size(); i++) {
        vector<float> v;
        for(int j = 0; j < filtered_annotations[i].data.size(); j++) {
            v.push_back(filtered_annotations[i].data[j][2]);
        }
        vector<bool> mask = mask1d_lt(v, this->keypoint_threshold);

        for(int k = 0; k < mask.size(); k++) {
            if(mask[k]) {
                for(int l = 0; l < filtered_annotations[i].data[0].size(); l++) {
                    filtered_annotations[i].data[k][l] = 0.0;
                }
            }
        }
    }

    vector<Annotation> filtered_annotations_by_score;
    for(int i = 0; i < filtered_annotations.size(); i++) {
        if(filtered_annotations[i].score() >= this->instance_threshold)
            filtered_annotations_by_score.push_back(filtered_annotations[i]);
    }

    sort(filtered_annotations_by_score.begin(), filtered_annotations_by_score.end(), compare_annotations);
    return filtered_annotations_by_score;
}